vpr/src/backend/physical_netlist_writer.h

```diff
@@ -0,0 +1,246 @@
+
+  public:
+    void visit_top_impl(const char* top_level_name) override;
+    void visit_route_through_impl(const t_pb* atom) override;
+    void visit_atom_impl(const t_pb* atom) override;
+    void visit_clb_impl(ClusterBlockId blk_id, const t_pb* clb) override;
+    void visit_all_impl(const t_pb_routes& top_pb_route, const t_pb* pb) override;
+    void finish_impl() override;
+
+  private:
+    vtr::string_internment* strings_;
+    Device::Reader& dr_;
+    PhysicalNetlist::PhysNetlist::Builder& phys_netlist_;
+
+    // String list for the physical netlist
+    StringEnumerator new_strings_;
+
+    IntraLbPbPinLookup pb_graph_pin_lookup_from_index_by_type_;
+    vtr::interned_string ic_primitive;
+    vtr::interned_string ic_routing;
+    vtr::interned_string ic_site_pip;
+    vtr::interned_string ic_generic;
+    vtr::interned_string ic_gridloc;
+
+    std::unordered_map<size_t, size_t> get_lut_pin_rotation(const t_pb* atom, bool is_lut);
+
+    // Propagates the internal pb type routing traversing all the routes
+    void iterate_nodes(std::vector<t_pb_graph_pin*>& pb_pins, bool is_source);
+
+    void walk_routing();
+    t_route_branch& walk_route_tree(const t_rt_node* root, t_route_branch& parent_branch);
+    void output_tree(t_route_branch& branch, PhysicalNetlist::PhysNetlist::RouteBranch::Builder builder, int level);
+
+    // TODO: For now consider only non-constant nets
+    std::unordered_map<size_t, t_route_branch> phys_nets_;
+
+    void build_netlist();
+
+    /* Helper containers and attributes */
+
+    int current_site_;
+    int current_site_type_;
+    bool is_pad_site_type_;
+    t_pb_routes curr_pb_route_;
+
+    t_physical_tile_type_ptr physical_tile_ = nullptr;
+    t_logical_block_type_ptr logical_block_ = nullptr;
+
+    std::unordered_set<int> site_pads_;
+
+    // Bel Cell mappings
+    std::unordered_map<uint32_t, std::set<t_bel_cell_mapping>> bel_cell_mappings_;
+
+    /// @brief vector of pair containing string index references of site instances
+    std::vector<std::pair<int, int>> site_instances_;
+
+    std::vector<t_cell_placement> cell_placements_;
+
+    /** @brief Returns the string corresponding to the given index */
+    std::string str(size_t idx) {
+        if ((int)idx == -1)
+            return std::string("constant_block");
+        return std::string(dr_.getStrList()[idx].cStr());
+    }
+
+    /** @brief Returns the bel name string index corresponding to its name and site type */
+    int get_bel(std::string name, int site_type_id) {
+        auto site_type = dr_.getSiteTypeList()[site_type_id];
+
+        for (auto bel : site_type.getBels())
+            if (str(bel.getName()) == name)
+                return bel.getName();
+
+        VTR_ASSERT(false);
+    }
+
+    /** @brief Returns true if the cell is of LUT type */
+    int is_lut(std::string name, int site_type_id) {
```

Comment:
Wrong type.
```suggestion
    bool is_lut(std::string name, int site_type_id) {
```

This functionality has already been implemented here: https://github.com/antmicro/vtr-verilog-to-routing/blob/4adcd0770a9b953b0e167bfabe535de0dccb9344/libs/libarchfpga/src/read_fpga_interchange_arch.cpp#L341

Instead of making multiple version of the same function, I believe we should generalize it and declare it in [libs/libarchfpga/src/fpga_interchange_arch_utils.h](https://github.com/antmicro/vtr-verilog-to-routing/blob/acom/fpga-interchange-rr-graph%2Bconstants/libs/libarchfpga/src/fpga_interchange_arch_utils.h)

Reply:
Right, moved to utils